File: shopping-cart/csharp/src/ShoppingCart/LineItem.cs
namespace ShoppingCart;

public sealed class LineItem
{
    public Product Product { get; }
    public Quantity Quantity { get; private set; }

    public LineItem(Product product, Quantity quantity)
    {
        Product = product;
        Quantity = quantity;
    }

    public Money Subtotal() => Product.DiscountPolicy.Apply(Product.UnitPrice, Quantity);

    internal void IncrementBy(Quantity delta)
    {
        Quantity = new Quantity(Quantity.Value + delta.Value);
    }

    internal void SetQuantity(Quantity quantity)
    {
        Quantity = quantity;
    }
}

File: shopping-cart/typescript/src/LineItem.ts
import { Money } from './Money.js';
import { Product } from './Product.js';
import { Quantity } from './Quantity.js';

export class LineItem {
  private _quantity: Quantity;

  constructor(public readonly product: Product, quantity: Quantity) {
    this._quantity = quantity;
  }

  get quantity(): Quantity { return this._quantity; }

  subtotal(): Money {
    return this.product.discountPolicy.apply(this.product.unitPrice, this._quantity);
  }

  incrementBy(delta: Quantity): void {
    this._quantity = this._quantity.plus(delta);
  }

  setQuantity(quantity: Quantity): void {
    this._quantity = quantity;
  }
}
